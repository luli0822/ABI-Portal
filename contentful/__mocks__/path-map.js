module.exports = {
	pathMap: {
		'/': {
			page: '/',
			query: {
				lang: 'en-CA',
				pageId: 'site-home-page',
			},
		},
		'/home-pg-fr': {
			page: '/',
			query: {
				lang: 'fr-CA',
				pageId: 'site-home-page',
			},
		},
		'/about-pg-en': {
			page: '/',
			query: {
				lang: 'en-CA',
				pageId: 'about-us-page',
			},
		},
		'/about-pg-fr': {
			page: '/',
			query: {
				lang: 'fr-CA',
				pageId: 'about-us-page',
			},
		},
		// This path map for testing refreshed index page
		'/refreshed-page': {
			page: '/',
			query: {
				lang: 'en-CA',
				pageId: 'REFRESHED-site-home-page',
			},
		},
	},
};
