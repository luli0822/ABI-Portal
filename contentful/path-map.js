// eslint-disable-next-line import/no-unresolved
const siteData = require('../data/siteData.json');
const { fallbackLang, defaultLang } = require('./config');
const { assetPrefix } = require('../site-config');
const { slugify } = require('../utils');

// this builds paths for a site that has both English and French pages.
// If you are building two separate sites for EN and FR, your path map will be very different
// Instead, you will use the language for the current context to build only pages for that language

function pathsForPageEntry(pageReference, isIndex = false) {
	const pageEntry = siteData.linkedEntries[pageReference.sys.id];
	const slugLanguages = Object.keys(pageEntry.fields.slug);
	// todo: map from template content model id or displayAs field
	const template = '';
	return slugLanguages.map(lang => {
		// if this is the index page for the default language, ignore the slug value
		const slug = isIndex && lang === defaultLang ? '' : pageEntry.fields.slug[lang];
		// Google Tag Manager(GTM) pageName. if GTM field is undefinded, use the page title as fallback
		const analyticsPageName = slugify(
			(pageEntry.fields.analyticsPageName && pageEntry.fields.analyticsPageName[fallbackLang]) ||
				(pageEntry.fields.title && pageEntry.fields.title[fallbackLang])
		);
		return {
			[`${assetPrefix}/${slug}`]: {
				// this "page" field maps to a template in the /pages directory
				page: `/${template}`,
				// lang and pageId will be passed to getInitialProps
				query: {
					lang,
					pageId: pageEntry.sys.id,
					analyticsPageName,
				},
			},
		};
	});
}

function pathMap() {
	const { siteEntry } = siteData;

	const indexPageEntry = siteEntry.fields.indexPage[fallbackLang];
	const pageEntries = siteEntry.fields.pages ? siteEntry.fields.pages[fallbackLang] : [];
	const indexPaths = pathsForPageEntry(indexPageEntry, true);
	const pagePaths = pageEntries.map(entry => pathsForPageEntry(entry));
	const allPaths = [indexPaths, ...pagePaths]
		.reduce((acc, val) => acc.concat(val), [])
		.reduce((acc, val) => ({ ...acc, ...val }), {});

	return allPaths;
}

module.exports = {
	// TODO: refactor to just export the function?
	// consider how to cache pathMap
	pathMap: pathMap(),
	pathsForPageEntry,
};
