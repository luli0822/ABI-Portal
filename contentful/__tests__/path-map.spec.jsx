import mockSiteData from '../../data/__mocks__/siteData.json';

jest.mock('../../data/siteData.json', () => ({ ...mockSiteData }), { virtual: true });

jest.mock('../config');

jest.mock('../../site-config.js');

// TODO: refactor path-map to export the function
// eslint-disable-next-line import/first
import { pathMap, pathsForPageEntry } from '../path-map';

describe('pathMap()', () => {
	it('generates the right paths from staticSiteData', async () => {
		expect(pathMap).toMatchSnapshot();
	});
});

describe('pathsForPageEntry()', () => {
	it('generates the EN and FR links for a page', async () => {
		// get the first page from the mock site data pages
		const reference = mockSiteData.siteEntry.fields.pages['en-CA'][0];
		const paths = pathsForPageEntry(reference);
		expect(paths).toMatchSnapshot();
	});

	it('generates the EN and FR links for an index page', async () => {
		// get the index page from the mock site data
		const reference = mockSiteData.siteEntry.fields.indexPage['en-CA'];
		const paths = pathsForPageEntry(reference, true);
		expect(paths).toMatchSnapshot();
	});
});
