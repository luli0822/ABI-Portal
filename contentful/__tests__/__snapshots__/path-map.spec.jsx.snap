// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pathMap() generates the right paths from staticSiteData 1`] = `
Object {
  "/some-asset-prefix/": Object {
    "page": "/",
    "query": Object {
      "analyticsPageName": "title-en",
      "lang": "en-CA",
      "pageId": "site-home-page",
    },
  },
  "/some-asset-prefix/about-pg-en": Object {
    "page": "/",
    "query": Object {
      "analyticsPageName": "title-en",
      "lang": "en-CA",
      "pageId": "about-us-page",
    },
  },
  "/some-asset-prefix/about-pg-fr": Object {
    "page": "/",
    "query": Object {
      "analyticsPageName": "title-en",
      "lang": "fr-CA",
      "pageId": "about-us-page",
    },
  },
  "/some-asset-prefix/home-pg-fr": Object {
    "page": "/",
    "query": Object {
      "analyticsPageName": "title-en",
      "lang": "fr-CA",
      "pageId": "site-home-page",
    },
  },
}
`;

exports[`pathsForPageEntry() generates the EN and FR links for a page 1`] = `
Array [
  Object {
    "/some-asset-prefix/about-pg-en": Object {
      "page": "/",
      "query": Object {
        "analyticsPageName": "title-en",
        "lang": "en-CA",
        "pageId": "about-us-page",
      },
    },
  },
  Object {
    "/some-asset-prefix/about-pg-fr": Object {
      "page": "/",
      "query": Object {
        "analyticsPageName": "title-en",
        "lang": "fr-CA",
        "pageId": "about-us-page",
      },
    },
  },
]
`;

exports[`pathsForPageEntry() generates the EN and FR links for an index page 1`] = `
Array [
  Object {
    "/some-asset-prefix/": Object {
      "page": "/",
      "query": Object {
        "analyticsPageName": "title-en",
        "lang": "en-CA",
        "pageId": "site-home-page",
      },
    },
  },
  Object {
    "/some-asset-prefix/home-pg-fr": Object {
      "page": "/",
      "query": Object {
        "analyticsPageName": "title-en",
        "lang": "fr-CA",
        "pageId": "site-home-page",
      },
    },
  },
]
`;
